resize.man
.\" $TOG: resize.man /main/14 1998/02/09 14:17:59 kaleb $
.TH RESIZE 1 "Release 6.4" "X Version 11"
.SH NAME
resize \- set TERMCAP and terminal settings to current xterm window size
.SH SYNOPSIS
.B resize
[ \fB\-u\fP | \fB\-c\fP ] [ \fB\-s\fP [ \fIrow col\fP ] ]
.SH DESCRIPTION
.I Resize
prints a shell command for setting the TERM and TERMCAP environment variables
to indicate the current size of \fIxterm\fP window from which the command
is run.  For this output to take effect, \fIresize\fP must either be evaluated
as part of the command line (usually done with a shell alias or function) or
else redirected to a file which can then be read in.  From the C shell (usually
known as \fI/bin/csh\fP), the following alias could be defined in the 
user's \fI.cshrc\fP:
.sp
.nf
        %  alias rs 'set noglob; eval \fC`\fPresize\fC`\fP'
.fi
.sp
After resizing the window, the user would type:
.sp
.nf
        %  rs
.fi
.sp
Users of versions of the Bourne shell (usually known as \fI/bin/sh\fP) that 
don't have command
functions will need to send the output to a temporary file and the read it back
in with the ``.'' command:
.sp
.nf
        $  resize > /tmp/out
        $  .\0/tmp/out
.fi
.SH OPTIONS
The following options may be used with \fIresize\fP:
.TP 8
.B \-u
This option indicates that Bourne shell commands should be generated even if 
the user's current shell isn't \fI/bin/sh\fP.
.TP 8
.B \-c
This option indicates that C shell commands should be generated even if the
user's current shell isn't \fI/bin/csh\fP.
.TP 8
.B \-s \fR[\fIrows columns\fP]
This option indicates that Sun console escape sequences will be used 
instead of the special \fIxterm\fP escape code.  If \fIrows\fP and
\fIcolumns\fP are given, \fIresize\fP will ask the \fIxterm\fP to resize
itself.  However, the window manager may choose to disallow the change.
.SH FILES
.TP 15
/etc/termcap
for the base termcap entry to modify.
.TP 15
~/.cshrc
user's alias for the command.
.SH "SEE ALSO"
csh(1), tset(1), xterm(1)
.SH AUTHORS
Mark Vandevoorde (MIT-Athena), Edward Moy (Berkeley)
.br
Copyright 1984, 1985 by The Open Group
.br
See
.IR X (1)
for a complete copyright notice.
.SH BUGS
The \fB\-u\fP or \fB\-c\fP must appear to the left of \fB\-s\fP if both are 
specified.


CHANGES
VERSION X10/6.6A
------- -------

General
-------

vt102 and Tektronix 4015 emulation in separate windows, each with its own
mode menu (control middle button).

Switching between each mode can be done via the menus, or with escape
sequences:
	vt102 -- \E[?38h --> tek
	tek   -- \E\003  --> vt102

The windows hilite themselves when the mouse is with either window or when
one of these is the focus window.  The windows unhilite when the mouse moves
out of the window and neither is the focus window.  Hiliting appears as
windows with solid color borders and solid block cursors.  Unhiliting appears
as windows with "grayed" borders and a rectangle cursor.

The parser for both modes is table-driven and is much more efficient than the
original parsers.

If UTMP is define in the Makefile, then each instance of xterm makes an entry
in /etc/utmp and thus appears with the "who" command.

If both windows are visible and one is iconified, both disappear.  Deiconifing
causes both to appear again.

If the initial mode is vt102, /etc/termcap is searched, in order, for the
following termcaps, until one is found:
	xterms, xterm, vt102, vt100, ansi, dumb
If the vt102 window happens to be 80x24, then xterms is added to the begining
of the list.  If the initial mode is tek, the termcaps searched for are:
	tek4015, tek4014, tek4013, tek4010, dumb

The window size is passed to the operating system via TIOCSWINSZ (4.3) or
TIOCSSIZE (sun).  A SIGWINCH siganl is sent if the vt102 window is resized.

For login windows, menus are disabled until a user has successfully login.

A new menu (xterm) has been added (control left button) that gives the version
number.  The menu allows titlebars to be displayed (see below), the windows
redrawn, or for various signals to be sent to the process group of the command
xterm is executing.  The signals include SIGCONT, SIGTSTP,SIGHUP, SIGINT,
SIGTERM and SIGKILL.

Titlebars for the window can be turned off with .Xdefaults
"xterm.TitleBar off", with command line option -tb or via the xterm menu.

The titlebar font is, by default, vtsingle.  The font can be changed with
.Xdefaults "xterm.TitleFont helv12b" for helv12b font or via command line
option -ft helv12b.

The titlebar look like Macintosh titlebars.  A series of strips appears when the
window is hilited, and are gone when unhilited.

Both windows are visible and the mouse cursor is in either window, only the
active window is hilited (before both windows were hilited).

A time delay has been added for window hiliting/unhiliting so that the mouse
passing momentarily through a window will not cause it to hilite and then
unhilite when autoraise is on.

Automatic raising of windows has been added, so that the window is automatically
raised (after the time delay) when the mouse enters a window.  The mode is
turned on via .Xdefaults "xterm.AutoRaise on", via command line option -ar
or via the xterm menu.  This is a dubious feature.

Visual bell mode can be turned on via .Xdefaults "xterm.VisualBell on", via
command line option -vb and from the xterm menu.

If the NOWINDOWMENU is not defined, then
Control right button activates a windows menu that lists the visible (named)
windows in back to front order reading down.  Selecting a window name cause
that window to be raised to the top (useful for getting to a buried
window).  This feature does not belong in xterm, and will be
completely removed in V11 of X, where window managers can be more intellegent.

The tek window has the same name as the vt102 window, but with " (Tek)"
appended (before they were the same).

During startup, xterm tries to create unique window names so that the
windows menu will not contain ambiguous windows.  It does this by look
through the list of windows and if a name exist that is the same, " #N"
will be appended to make it unique (N is a number two or greater).

Some reorganization of the code and data has been done to put more data in
the text segment so many xterm processes can share the (read-only) data, rather
than having n copies in core.

There are several bug fixes including better sun compatibility and a problem
in the vt102 parse table.

With some mods to libX.a, xterm solves a problem of running xterm on a remote
host with a local window, when the remote host uses a different keymap (like
running on a remote microvax on a sun).  If the environment variable KEYBD
is set or the command line option "-kb name" is given, then a keymap file
with the given name is used, rather than the default of your .Xkeymap file.
Standard names for keymap files is yet to be made.  (This option is turned on
via defining KEYBD in the Makefile.)

Xterm now always creates its own icon(s), which includes the window's name.
The -i option now means startup showing the icon.  The option #[+-]x[+-]y
means to position the icon there instead of centered within the window.
The bitmap within the icon show the current active mode (either vt102 or
tek).  The icon reverse-videos with the windows.
By default, the icons are miniture versions of the window.

When the mouse is in the stripped area of the titlebar, you can get any
of the three menus without pressing the control key.  Clicking in the
actual title iconifies.  Clicking in an icon deiconifies it.

Many mode option on the command line allow a '+' instead of '-' to indicate
turning off the indicated mode.

Color support has been improved, so that an unselected cursor is an empty
rectangle with the foreground color, and the selected cursor is a filled
rectangle with the cursor color.

A better and faster menu package has been added that supports (Mac-like)
checkmarks and dimmed (disabled and grayed) item, as well as line separator
items.

To conserve on resources, the unhilited border of the window(s) is now
always a gray pixmap, which is shared by the menu package.

Logging capability has been added, in which all characters sent to the
xterm window are recorded.  The mode is turned on by using .Xdefaults option

resize.c
#include <X/mit-copyright.h>

/* Copyright    Massachusetts Institute of Technology    1984	*/

/* resize.c */

#include <stdio.h>
#include <sgtty.h>
#include <strings.h>
#include <sys/ioctl.h>

#ifndef lint
static char *rcsid_resize_c = "@Header: resize.c,v 10.5 86/02/01 16:06:55 tony Rel @";
#endif

char *myname;

char *strindex (), *index (), *rindex();

main (argc, argv)
char **argv;
/*
   resets termcap string to reflect current screen size
 */
{
	int rows, cols;
	char *size = "\0337\033[r\033[999;999H\033[6n";
	struct sgttyb sg, sgorig;
	char termcap [1024];
	char newtc [1024];
	char *setname = "";
	register char *ptr, *env;
	char *getenv();
#ifdef TIOCSWINSZ
	struct winsize ws;
#endif
	tgetent (termcap, "xterm");

	if(ptr = rindex(myname = argv[0], '/'))
		myname = ptr + 1;
	if((env = getenv("TERMCAP")) && *env)
		strcpy(termcap, env);
	else {
		if(!(env = getenv("TERM")) || !*env) {
			env = "xterm";
			setname = "setenv TERM xterm;\n";
		}
		if(tgetent (termcap, env) <= 0) {
			fprintf(stderr, "%s: Can't get entry \"%s\"\n",
			 myname, env);
			exit(1);
		}
	}

 	ioctl ( 0, TIOCGETP, &sgorig);
	sg = sgorig;
	sg.sg_flags |= RAW;
	ioctl ( 0, TIOCSETP, &sg);

	write ( 0, size, strlen (size));
	scanf ("\033[%d;%dR", &rows, &cols);
	write ( 0, "\0338", 2);

	ioctl ( 0, TIOCSETP, &sgorig);

	/* update termcap string */
	/* first do columns */
	if ((ptr = strindex (termcap, "co#")) == NULL) {
		fprintf(stderr, "%s: No `co#'\n", myname);
		exit (1);
	}
	strncpy (newtc, termcap, ptr - termcap + 3);
	sprintf (newtc + strlen (newtc), "%d", cols);
	ptr = index (ptr, ':');
	strcat (newtc, ptr);

	/* now do lines */
	if ((ptr = strindex (newtc, "li#")) == NULL) {
		fprintf(stderr, "%s: No `li#'\n", myname);
		exit (1);
	}
	strncpy (termcap, newtc, ptr - newtc + 3);
	sprintf (termcap + ((int) ptr - (int) newtc + 3), "%d", rows);
	ptr = index (ptr, ':');
	strcat (termcap, ptr);

	printf ("set noglob;\n%ssetenv TERMCAP '%s';\nunset noglob;\n",
	 setname, termcap);
#ifdef TIOCGWINSZ
	/* finally, set the tty's window size */
	if (ioctl (0, TIOCGWINSZ, &ws) != -1) {
	    /* we don't have any way of directly finding out
	       the current height & width of the window in pixels.  We try
	       our best by computing the font height and width from the "old"
	       struct winsize values, and multiplying by these ratios...*/
	    if (ws.ws_xpixel != 0)
	        ws.ws_xpixel = cols*(ws.ws_xpixel/ws.ws_col);
	    if (ws.ws_ypixel != 0)
	        ws.ws_ypixel = rows*(ws.ws_ypixel/ws.ws_row);
	    ws.ws_row = rows;
	    ws.ws_col = cols;
	    ioctl (0, TIOCSWINSZ, &ws);
	    }
#endif
	exit(0);
}

char *strindex (s1, s2)
/*
   returns a pointer to the first occurrence of s2 in s1, or NULL if there are
   none.
 */
char *s1, *s2;
{
	char *s3;

	while ((s3 = index (s1, *s2)) != NULL)
	{
		if (strncmp (s3, s2, strlen (s2)) == 0) return (s3);
		s1 = ++s3;
	}
	return (NULL);
}

